### PR TITLE
✨ Introduce controllerutil.Object interface

### PR DESCRIPTION
**Pull request by ruromero**
Signed-off-by: ruromero <rromerom@redhat.com>

Introduce an interface that combines `metav1.Object` and `runtime.Object` as discussed in https://github.com/kubernetes-sigs/controller-runtime/issues/594

The main purpose of this interface is to reduce type conversions between the two and provide generic method signatures that can show potential errors at compile time and help developers to write simpler code.

This can make reconsider many places in this library and dependent ones its usage each time a cast is done, usually like the following:
```
//metav1.Object -> runtime.Object
ro, ok := owner.(runtime.Object)

//runtime.Object -> metav1.Object
o, err := meta.Accessor(obj)
```
An example where this interface can be used is https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/controllerutil.go#L95

### CODE REVIEW

**Comment by ruromero** [comment]
Added CLA

File: pkg/client/interfaces.go [hunk 1]
```diff
@@ -133,3 +134,10 @@ func IgnoreNotFound(err error) error {
 	}
 	return err
 }
+
+// KubernetesResource allows functions to work indistinctly with any resource that
+// implements both Object interfaces.
+type KubernetesResource interface {
```
Comment (joelanford) [inline pkg/client/interfaces.go:140]: Given that this is a union of two `*.Object` types would it be more intuitive for this to also include `Object` in the name?

Also, I wonder if the `client` package is the most appropriate place for this type. Nothing is really jumping out to me as an alternate though. Perhaps `pkg/controller/controllerutil`? :man_shrugging:

Comment (ruromero) [reply pkg/client/interfaces.go]: Renamed and moved

**Comment by DirectXMan12** [comment]
> decide whether we want to update those function signatures as well.

Yeah, I think we want to do that.  I *think* there's actually situations where some of those aren't a breaking change (I seem to recall adding additional interfaces to a compound interface casts transparently or something, but I'd have to double check).

**Comment by vincepri** [comment]
Given that this lives under `controllerutil`, could we just call it `Object` instead of `KubernetesObject `?

**Comment by ruromero** [comment]
renamed @vincepri

**Comment by vincepri** [comment]
Should we backport this? I'd be useful for other projects as well

**Comment by luxas** [comment]
Thanks for this change :100:!
I'd love to have this under core `k8s.io/apimachinery`.
Would someone be up for "backporting" this to there?

**Comment by vincepri** [comment]
@luxas Happy to help, although need some pointers to where we should place this interface :)

**Comment by luxas** [comment]
Without knowing exactly, maybe https://pkg.go.dev/k8s.io/apimachinery@v0.18.5/pkg/api/meta?tab=doc would be a good place?

**Pull request merged**